flatfs.go

```diff
@@ -179,7 +181,14 @@ type opResult struct {
 	name  string
 }
 
-// Returns nil if there's nothing to do.
+// Begins starts the processing of an op:
+// - if no other op for the same key exist, register it and return immediately
+// - if another op exist for the same key, wait until it's done:
+// 		- if that previous op succeeded, consider that ours shouldn't execute and return nil
+// 		- if that previous op failed, start ours
+//
+// TODO: honestly those rules are weird to me and might lead to data loss? Why a later concurrent op
+//  should not execute?
```

Comment:
I suspect this has to do with flatfs not really being used as a generic datastore but as a backing for a blockstore. This means that (key, value) pairs are immutable so what's your other operation going to do other than a) write the data that's already present b) delete the data you just wrote which is likely due to an application layer bug.

I suspect this isn't the only place where we have these assumptions another is related to the domain of valid key names.

Reply:
Something non-obvious when reading from this PR: this synch/dedup structure is shared by all write functions of the datastore, so a `Delete` could be cancelled by a recent `Put` for example. That would be a very legal scenario, even if attached to a blockstore. My point is that this code brings a lot of complexity for a rare case (dedup writes on the same key) for little benefit.